### PR TITLE
Bulk load CDK: add complex type tests

### PR DESCRIPTION
**Pull request by edgao**
port over DAT.testProblematicTypes to the new framework. I split this into two separate tests:
* testContainerTypes has object/arrays (with/without schema)
* testUnions does all the union stuff
* I discarded a few parts of the test that were redundant (mostly b/c we have stronger schema handling now, so some schemas would have been duplicates)

There were a few behavior changes / bugfixes in other parts of the code; see the comments for explanation.

### CODE REVIEW

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValue.kt [hunk 1]
```diff
@@ -148,7 +148,7 @@ value class TimeValue(val value: String) : AirbyteValue, Comparable<TimeValue> {
 @JvmInline
 value class ArrayValue(val values: List<AirbyteValue>) : AirbyteValue {
     companion object {
-        fun from(list: List<Any?>): ArrayValue = ArrayValue(list.map { it as AirbyteValue })
+        fun from(list: List<Any?>): ArrayValue = ArrayValue(list.map { AirbyteValue.from(it) })
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValue.kt:151]: this was a brain fart, simple bugfix

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt [hunk 2]
```diff
@@ -40,10 +40,14 @@ class JsonToAirbyteValue {
                 is TimestampTypeWithTimezone,
                 is TimestampTypeWithoutTimezone -> TimestampValue(json.asText())
                 is UnionType -> toUnion(json, schema.options)
-                is UnknownType -> UnknownValue("From $schema: $json")
+                // If we fail to recognize the schema, just pass the json through directly.
+                // This enables us to more easily add new types, without breaking compatibility
+                // within existing connections.
+                is UnknownType -> fromJson(json)
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt:46]: @johnny-schmidt you possibly have this fixed in one of your open PRs? I actually don't remember 😅

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt [hunk 3]
```diff
@@ -103,18 +107,13 @@ class JsonToAirbyteValue {
         if (!json.isObject) {
             throw IllegalArgumentException("Could not convert $json to Object")
         }
+        val objectProperties = LinkedHashMap<String, AirbyteValue>()
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt:110]: what we discussed earlier in the week - pass through all object properties instead of filtering

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/test/util/RecordDiffer.kt [hunk 4]
```diff
@@ -297,8 +297,9 @@ class RecordDiffer(
                     }
                     // otherwise, just be a terrible person.
                     // we know these are the same type, so this is safe to do.
-                    else ->
+                    is Comparable<*> ->
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/test/util/RecordDiffer.kt:300]: this continues to be the hackiest part of the record differ - object/array aren't comparable

File: airbyte-cdk/bulk/toolkits/load-avro/src/main/kotlin/io/airbyte/cdk/load/data/avro/AirbyteValueToAvroRecord.kt [hunk 5]
```diff
@@ -27,7 +27,9 @@ class AirbyteValueToAvroRecord {
             is ObjectValue -> {
                 val record = GenericData.Record(schema)
                 airbyteValue.values.forEach { (name, value) ->
-                    record.put(name, convert(value, schema.getField(name).schema()))
+                    schema.getField(name)?.let { field ->
```
Comment (edgao) [inline airbyte-cdk/bulk/toolkits/load-avro/src/main/kotlin/io/airbyte/cdk/load/data/avro/AirbyteValueToAvroRecord.kt:30]: also what we talked about - avro needs to explicitly filter down to declared fields

File: airbyte-cdk/bulk/toolkits/load-csv/src/testFixtures/kotlin/io/airbyte/cdk/load/data/csv/CsvRowToAirbyteValue.kt [hunk 6]
```diff
@@ -63,7 +64,8 @@ class CsvRowToAirbyteValue {
                     .fields()
                     .asSequence()
                     .map { entry ->
-                        entry.key to entry.value.toAirbyteValue(field.properties[entry.key]!!.type)
+                        val type = field.properties[entry.key]?.type ?: UnknownType("unknown")
+                        entry.key to entry.value.toAirbyteValue(type)
```
Comment (edgao) [inline airbyte-cdk/bulk/toolkits/load-csv/src/testFixtures/kotlin/io/airbyte/cdk/load/data/csv/CsvRowToAirbyteValue.kt:68]: csv needs (?) to explicitly pass through undeclared fields, so we just default to an unknown type

File: airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt [hunk 7]
```diff
@@ -47,21 +55,126 @@ abstract class S3V2WriteTest(path: String) :
...
+    }
+
+    @Test
+    override fun testUnions() {
+        super.testUnions()
+    }
 }
 
-class S3V2WriteTestJsonUncompressed : S3V2WriteTest(S3V2TestUtils.JSON_UNCOMPRESSED_CONFIG_PATH)
+class S3V2WriteTestJsonUncompressed :
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt:70]: tl;dr:
* json+csv don't stringify schemaless objects, don't promote unions to objects, and do preserve undeclared fields
* avro+parquet don't stringify, and don't preserve undeclared fields
    * avro doesn't promote unions
    * parquet does promote unions

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/write/BasicFunctionalityIntegrationTest.kt [hunk 8]
```diff
@@ -841,6 +852,426 @@ abstract class BasicFunctionalityIntegrationTest(
         )
     }
 
+    // TODO basic allTypes() test
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/write/BasicFunctionalityIntegrationTest.kt:855]: this will be a separate PR

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt (hunk 2 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt:46]: I don't think so

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt [hunk 9]
```diff
@@ -69,6 +69,7 @@ class JsonFormattingWriter(
 ) : ObjectStorageFormattingWriter {
     override fun accept(record: DestinationRecord) {
         outputStream.write(recordDecorator.decorate(record).toJson().serializeToString())
+        outputStream.write("\n")
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:72]: 😊

**Review by johnny-schmidt — APPROVED** [review]
Great! Thanks for following through on all the little fixes!
